fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/OCFLPersistentStorageSession.java

```diff
@@ -177,6 +187,51 @@ public void persist(final ResourceOperation operation) throws PersistentStorageE
 
     }
 
+    private FedoraOCFLMapping findOrCreateFedoraOCFLMapping(final ResourceOperation operation, final String resourceId)
+            throws PersistentStorageException {
+        FedoraOCFLMapping mapping;
+
+        try {
+            mapping = fedoraOcflIndex.getMapping(resourceId);
+        } catch (FedoraOCFLMappingNotFoundException e) {
+            //if no mapping exists, create one
+            if (operation instanceof CreateResourceOperation) {
```

Comment:
I'm not sure at the moment how else to arrange things, but it feels a little odd to me to have management of the index being so directly controlled by this session. There will be other places that add to the index, at a minimum during a rebuild scenario. It might make sense for the index to know how to populate itself based off what's on disk, so a call to `getMapping` would seek an existing record first in the index and then on disk, before failing. I know this is first pass implementation, just wondering if more of the guts should be internalized.

Reply:
I'll take a look and see what makes sense.

Reply:
I agree that OCFLPersistentStorageSession should not be responsible for populating the index when the resources already exist on disk and that yes, maybe some kind of lazy loading would be a good idea ( and also we could support preloading the index).   Still I believe we need a way to populate the index when new objects are created no?

Reply:
I think its reasonable for the index to have both `getMapping` and `getOrCreateMapping` where the latter attempts to lazy load the value into the index by whatever means it feels is appropriate.

so `getOrCreateMapping` would:

1. Check to see if the resource id existed in the index
2. if not, then check to see if the resource id was persisted as a OCFL object. So given `atomic/nested_ag` it would try to see if the ocfl root contained an object with that id. This covers the "new root resource" case.
3. if not, then it would remove the last segment of the identifier to get the parent and check for it in the index. If it finds an entry for that parent, then the parent's root resource would become the current resource's root resource. So given `my_ag/folder/binary.txt`, it would look up `my_ag/folder`, then use its root resource which is `my_ag`

The parent should not be new at the same time as the new resource, so it should be indexed (unless the index got wiped, in which cause it should get repopulated at startup of the server).

I think the current approach might be problematic since the operation's parent id is the resource the contains the new resource, which is not necessarily the root resource (in the case of being contained by an atomistic resource).

Reply:
In step 2,  how would the OCFL Id be resolved in order to make that check?

Reply:
Is the OCFL id a function of the fedora id, or is it totally independent? If its totally independent, then it can only be found with a separate index of OCFL object ids or a crawl of all OCFL objects in the ocfl root.

Reply:
I think the answer is both:  it can be deterministic or independent depending on the object object minter. I believe we know we'll need to support independent ids.